src/main/scala/cognite/spark/v1/FlexibleDataModelRelationUtils.scala

```diff
@@ -15,58 +14,81 @@ import com.cognite.sdk.scala.v1.fdm.common.properties.PropertyType.{
   PrimitiveProperty,
   TextProperty
 }
+import com.cognite.sdk.scala.v1.fdm.common.properties.{PrimitivePropType, PropertyDefinition}
 import com.cognite.sdk.scala.v1.fdm.common.sources.SourceReference
+import com.cognite.sdk.scala.v1.fdm.instances.InstanceDeletionRequest.{
+  EdgeDeletionRequest,
+  NodeDeletionRequest
+}
 import com.cognite.sdk.scala.v1.fdm.instances.NodeOrEdgeCreate.{EdgeWrite, NodeWrite}
-import com.cognite.sdk.scala.v1.fdm.instances.{EdgeOrNodeData, InstancePropertyValue, NodeOrEdgeCreate}
+import com.cognite.sdk.scala.v1.fdm.instances.{
+  EdgeOrNodeData,
+  InstanceDeletionRequest,
+  InstancePropertyValue,
+  NodeOrEdgeCreate
+}
 import io.circe.syntax.EncoderOps
 import org.apache.spark.sql.Row
 import org.apache.spark.sql.types.StructType
 
 import java.time._
 import scala.util.{Failure, Success, Try}
 
-// scalastyle:off number.of.methods
+// scalastyle:off number.of.methods file.size.limit
 object FlexibleDataModelRelationUtils {
   private[spark] def createNodes(
-      instanceSpaceExternalId: String,
+      instanceSpace: Option[String],
       rows: Seq[Row],
       schema: StructType,
       propertyDefMap: Map[String, ViewPropertyDefinition],
       source: SourceReference): Either[CdfSparkException, Vector[NodeWrite]] =
     validateRowFieldsWithPropertyDefinitions(schema, propertyDefMap) *> createNodeWriteData(
-      instanceSpaceExternalId,
+      instanceSpace,
       schema,
       source,
       propertyDefMap,
       rows)
 
+  private[spark] def createNodes(
+      instanceSpace: Option[String],
+      rows: Seq[Row],
+      schema: StructType): Either[CdfSparkException, Vector[NodeWrite]] =
+    createNodeWriteData(instanceSpace, schema, rows)
+
   private[spark] def createEdges(
-      instanceSpaceExternalId: String,
+      instanceSpace: Option[String],
       rows: Seq[Row],
       schema: StructType,
       propertyDefMap: Map[String, ViewPropertyDefinition],
       source: SourceReference): Either[CdfSparkException, Vector[EdgeWrite]] =
     validateRowFieldsWithPropertyDefinitions(schema, propertyDefMap) *> createEdgeWriteData(
-      instanceSpaceExternalId,
+      instanceSpace,
       schema,
       source,
       propertyDefMap,
       rows)
 
+  private[spark] def createEdges(
+      instanceSpace: Option[String],
+      rows: Seq[Row],
+      schema: StructType): Either[CdfSparkException, Vector[EdgeWrite]] =
+    createEdgeWriteData(instanceSpace, schema, rows)
+
   private[spark] def createNodesOrEdges(
-      instanceSpaceExternalId: String,
+      instanceSpace: Option[String],
       rows: Seq[Row],
       schema: StructType,
       propertyDefMap: Map[String, ViewPropertyDefinition],
       source: SourceReference
   ): Either[CdfSparkException, Vector[NodeOrEdgeCreate]] =
     rows.toVector.traverse { row =>
       for {
+        space <- instanceSpace.map(Right(_)).getOrElse(extractSpaceExternalId(schema, row))
```

Comment:
`instanceSpace` is optional for the `Relation` and therefore `space` will be populated from data.